src/Microsoft.Diagnostics.Monitoring.WebApi/Stacks/EventStacksPipeline.cs

```diff
@@ -0,0 +1,153 @@
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Monitoring.EventPipe;
+using Microsoft.Diagnostics.NETCore.Client;
+using Microsoft.Diagnostics.Tracing;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.WebApi.Stacks
+{
+    internal sealed class EventStacksPipelineSettings : EventSourcePipelineSettings
+    {
+        public EventStacksPipelineSettings()
+        {
+            Duration = System.Threading.Timeout.InfiniteTimeSpan;
+        }
+
+        public TimeSpan Timeout { get; set; } = TimeSpan.FromSeconds(5);
+    }
+
+    internal sealed class EventStacksPipeline : EventSourcePipeline<EventStacksPipelineSettings>
+    {
+        private TaskCompletionSource<StackResult> _stackResult = new();
+        private StackResult _result = new();
+
+        public EventStacksPipeline(DiagnosticsClient client, EventStacksPipelineSettings settings)
+            : base(client, settings)
+        {
+        }
+
+        protected override MonitoringSourceConfiguration CreateConfiguration()
+        {
+            return new EventPipeProviderSourceConfiguration(requestRundown: false, bufferSizeInMB: 256, new[]
+            {
+                new EventPipeProvider(StackEvents.Provider, System.Diagnostics.Tracing.EventLevel.LogAlways)
+            });
+        }
+
+        protected override async Task OnEventSourceAvailable(EventPipeEventSource eventSource, Func<Task> stopSessionAsync, CancellationToken token)
+        {
+            eventSource.Dynamic.AddCallbackForProviderEvents((string provider, string _) => provider == StackEvents.Provider ?
+            EventFilterResponse.AcceptEvent : EventFilterResponse.RejectEvent, Callback);
+
+            using EventTaskSource<Action> sourceComplete = new EventTaskSource<Action>(
+                taskComplete => taskComplete,
+                addHandler => eventSource.Completed += addHandler,
+                removeHandler => eventSource.Completed -= removeHandler,
+                token);
+
+            // This is the same issue as GCDumps. We don't always get events back in realtime, so we have to stop the session and then process the events.
+            Task eventsTimeoutTask = Task.Delay(Settings.Timeout, token);
+            Task completedTask = await Task.WhenAny(_stackResult.Task, eventsTimeoutTask);
+
+            token.ThrowIfCancellationRequested();
+
+            await stopSessionAsync();
+            await sourceComplete.Task;
+
+            if (_stackResult.Task.Status != TaskStatus.RanToCompletion)
+            {
+                throw new InvalidOperationException("Unable to process stack in timely manner.");
+            }
+        }
+
+        public Task<StackResult> Result => _stackResult.Task;
+
+        private void Callback(TraceEvent action)
+        {
+            if (action.ProviderName != StackEvents.Provider)
```

Comment:
This check should not be necessary since the callback registration is already filtering this.